cluster/cluster.go

```diff
@@ -0,0 +1,437 @@
+	l log.Logger,
+	reg prometheus.Registerer,
+	bindAddr string,
+	advertiseAddr string,
+	knownPeers []string,
+	waitIfEmpty bool,
+	pushPullInterval time.Duration,
+	gossipInterval time.Duration,
+) (*Peer, error) {
+	bindHost, bindPortStr, err := net.SplitHostPort(bindAddr)
+	if err != nil {
+		return nil, err
+	}
+	bindPort, err := strconv.Atoi(bindPortStr)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid listen address")
+	}
+	var advertiseHost string
+	var advertisePort int
+
+	if advertiseAddr != "" {
+		var advertisePortStr string
+		advertiseHost, advertisePortStr, err = net.SplitHostPort(advertiseAddr)
+		if err != nil {
+			return nil, errors.Wrap(err, "invalid advertise address")
+		}
+		advertisePort, err = strconv.Atoi(advertisePortStr)
+		if err != nil {
+			return nil, errors.Wrap(err, "invalid advertise address, wrong port")
+		}
+	}
+
+	resolvedPeers, err := resolvePeers(context.Background(), knownPeers, advertiseAddr, net.Resolver{}, waitIfEmpty)
+	if err != nil {
+		return nil, errors.Wrap(err, "resolve peers")
+	}
+	level.Debug(l).Log("msg", "resolved peers to following addresses", "peers", strings.Join(resolvedPeers, ","))
+
+	// Initial validation of user-specified advertise address.
+	addr, err := calculateAdvertiseAddress(bindHost, advertiseHost)
+	if err != nil {
+		level.Warn(l).Log("err", "couldn't deduce an advertise address: "+err.Error())
+	} else if hasNonlocal(resolvedPeers) && isUnroutable(addr.String()) {
+		level.Warn(l).Log("err", "this node advertises itself on an unroutable address", "addr", addr.String())
+		level.Warn(l).Log("err", "this node will be unreachable in the cluster")
+		level.Warn(l).Log("err", "provide --cluster.advertise-address as a routable IP address or hostname")
+	}
+
+	// TODO(fabxc): generate human-readable but random names?
+	name, err := ulid.New(ulid.Now(), rand.New(rand.NewSource(time.Now().UnixNano())))
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Peer{
+		states: map[string]State{},
+		stopc:  make(chan struct{}),
+	}
+	p.delegate = newDelegate(l, reg, p)
+
+	cfg := memberlist.DefaultLANConfig()
+	cfg.Name = name.String()
+	cfg.BindAddr = bindHost
+	cfg.BindPort = bindPort
+	cfg.Delegate = p.delegate
+	cfg.Events = p.delegate
+	cfg.GossipInterval = gossipInterval
+	cfg.PushPullInterval = pushPullInterval
+	cfg.LogOutput = ioutil.Discard
+
+	if advertiseAddr != "" {
+		cfg.AdvertiseAddr = advertiseHost
+		cfg.AdvertisePort = advertisePort
+	}
+
+	ml, err := memberlist.Create(cfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "create memberlist")
+	}
+	p.mlist = ml
+
+	n, _ := ml.Join(knownPeers)
+	level.Debug(l).Log("msg", "joined cluster", "peers", n)
+
+	if n > 0 {
+		go p.warnIfAlone(l, 10*time.Second)
+	}
+	return p, nil
+}
+
+func (p *Peer) warnIfAlone(logger log.Logger, d time.Duration) {
+	tick := time.NewTicker(d)
+	defer tick.Stop()
+
+	for {
+		select {
+		case <-p.stopc:
+			return
+		case <-tick.C:
+			if n := p.mlist.NumMembers(); n <= 1 {
```

Comment:
I would argue this is a readiness condition if HA/gossiping is enabled

Reply:
Readyness in what sense? For a readyness endpoint, sure. Not for blocking actually though.
Since AM doesn't seem to have a readiness endpoint still, I guess that's for another PR.

Reply:
Just discussed this offline and decided that this is better served with a metric/alert, as otherwise we have a problem with the initial peer. Plus there is no readiness handler yet anyways.